pkg/build/registry/rest.go

```diff
@@ -18,41 +19,79 @@ var (
 	ErrBuildDeleted      = fmt.Errorf("build was deleted")
 )
 
+type ErrWatchError struct {
+	error
+}
+
 // WaitForRunningBuild waits until the specified build is no longer New or Pending. Returns true if
 // the build ran within timeout, false if it did not, and an error if any other error state occurred.
 // The last observed Build state is returned.
 func WaitForRunningBuild(buildClient buildtypedclient.BuildsGetter, build *buildapi.Build, timeout time.Duration) (*buildapi.Build, bool, error) {
 	fieldSelector := fields.OneTermEqualSelector("metadata.name", build.Name)
 	options := metav1.ListOptions{FieldSelector: fieldSelector.String(), ResourceVersion: build.ResourceVersion}
-	w, err := buildClient.Builds(build.Namespace).Watch(options)
-	if err != nil {
-		return build, false, err
-	}
 
-	observed := build
-	_, err = watch.Until(timeout, w, func(event watch.Event) (bool, error) {
-		obj, ok := event.Object.(*buildapi.Build)
-		if !ok {
-			return false, fmt.Errorf("received unknown object while watching for builds: %T", event.Object)
-		}
-		observed = obj
+	done := make(chan interface{}, 1)
+	defer close(done)
+	var resultBuild *buildapi.Build
+	var success bool
+	var resultErr error
 
-		if event.Type == watch.Deleted {
-			return false, ErrBuildDeleted
-		}
-		switch obj.Status.Phase {
-		case buildapi.BuildPhaseRunning, buildapi.BuildPhaseComplete, buildapi.BuildPhaseFailed, buildapi.BuildPhaseError, buildapi.BuildPhaseCancelled:
-			return true, nil
-		case buildapi.BuildPhaseNew, buildapi.BuildPhasePending:
-		default:
-			return false, ErrUnknownBuildPhase
+	deadline := time.Now().Add(timeout)
+	go func() {
+		for time.Now().Before(deadline) {
+			w, err := buildClient.Builds(build.Namespace).Watch(options)
```

Comment:
restarting the watch with the same options.ResourceVersion will continuously give you a 410 error if the resource version has expired